### PR TITLE
ASGI middleware/plugin.

### LINKED ISSUE #78: Honeybadger FastAPI plugin.
Hello!
I would love to add [FastAPI](https://fastapi.tiangolo.com/) support to the Honeybadger package, and I'd love to discuss some issues before diving deeper into the code.

First off, I'm already using Honeybadger with this framework (also in production): the solution I adopted is having a middleware handing the exceptions and notifying them when needed.

This approach would have the **advantage** of the possibility of implementing a single ASGI middleware for all the [ASGI-based frameworks](https://asgi.readthedocs.io/en/latest/implementations.html), but the **disadvantage** of not having an easy way for consuming the requests bodies in the middleware (and therefore sending them to Honeybadger). This is a [known issue](https://github.com/encode/starlette/issues/495#issuecomment-494008175). I've been having a look around and Sentry has adopted [the middleware solution](https://github.com/getsentry/sentry-python/blob/master/sentry_sdk/integrations/asgi.py), accepting the drawbacks of not parsing the body.

However, I understand that the request body may include useful piece of information from a debugging point of view. An alternative would be implementing and setting a [custom APIRoute](https://fastapi.tiangolo.com/advanced/custom-request-and-route), which is the neat workaround offered by FastAPI to give access to the body of the request, but it's coming with its own drawbacks aswell:

1. It would work just for FastAPI; (not a big issue)
2. What if a FastAPI's user wants to set his own custom APIRoute? (that's quite of an issue)
3. Users would need to set the custom APIRoute on their *app* *and* on each of their *routers*, or we'd have to elaborate some funny way of doing it under their backs, which doesn't sound too much of a good idea to me.

---

So, if that was me, I would keep the middleware solution and maybe offer the *custom APIRoute* as a "know-what-you-re-doing" alternative, but I would love an opinion on this approach.

Thank you!

### PR DESCRIPTION
**Pull request by stefanondisponibile**
Hello! I'm proposing the following PR that should close #78.

This PR adds ASGI applications support and a custom FastAPI route.
I added some tests and update the documentation.

What do you think @joshuap and @Kelvin4664?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -130,6 +130,99 @@ def lambda_handler(event, context):
 
     return (a/b) #This will be reported
 ```
+
+### [ASGI](https://asgi.readthedocs.io/en/latest/).
```
Comment (joshuap) [inline README.md:134]:
```suggestion
### ASGI
```

File: README.md [hunk 2]
```diff
@@ -130,6 +130,99 @@ def lambda_handler(event, context):
...
+```python
+from honeybadger import honeybadger, contrib
+
+honeybadger.configure(api_key='<your-api-key>')
+some_possibly_failing_function()  # you can track errors happening before your plugin initialization.
+asgi_application = someASGIApplication()
+asgi_application = contrib.ASGIHoneybadger(asgi_application)
+```
+
+### [FastAPI](https://fastapi.tiangolo.com/).
```
Comment (joshuap) [inline README.md:168]:
```suggestion
### FastAPI
```

File: README.md [hunk 3]
```diff
@@ -130,6 +130,99 @@ def lambda_handler(event, context):
...
+
+FastAPI is based on Starlette, a ASGI application.
```
Comment (joshuap) [inline README.md:170]:
```suggestion
[FastAPI](https://fastapi.tiangolo.com/) is based on Starlette, an ASGI application.
```

File: README.md [hunk 4]
```diff
@@ -130,6 +130,99 @@ def lambda_handler(event, context):
 
     return (a/b) #This will be reported
 ```
+
+### [ASGI](https://asgi.readthedocs.io/en/latest/).
+
+A generic ASGI middleware plugin is available for initializing and configuring Honeybadger: [`honeybadger.contrib.asgi`](./honeybadger/contrib/asgi.py).
```
Comment (joshuap) [inline README.md:136]:
```suggestion
A generic [ASGI](https://asgi.readthedocs.io/en/latest/) middleware plugin is available for initializing and configuring Honeybadger: [`honeybadger.contrib.asgi`](./honeybadger/contrib/asgi.py).
```

File: README.md [hunk 5]
```diff
@@ -130,6 +130,99 @@ def lambda_handler(event, context):
...
+The general pattern for these cases is wrapping your ASGI application with a middleware:
+
+```python
+from honeybadger import contrib
+
+asgi_application = someASGIApplication()
+asgi_application = contrib.ASGIHoneybadger(asgi_application)
+```
+
+You can pass configuration parameters (or *additional* configuration parameters) as keyworkd arguments at plugin's initialization:
```
Comment (joshuap) [inline README.md:147]:
```suggestion
You can pass configuration parameters (or *additional* configuration parameters) as keyword arguments at plugin's initialization:
```

File: README.md [hunk 6]
```diff
@@ -130,6 +130,99 @@ def lambda_handler(event, context):
...
+from honeybadger import honeybadger, contrib
+
+honeybadger.configure(api_key="<your-api-key>")
+app = FastAPI()
+app.add_middleware(contrib.ASGIHoneybadger, params_filters=["dont-include-this"])
+```
+
+#### FastAPI advanced usage.
+
+Consuming the request body in a ASGI application's middleware is [problematic and discouraged](https://github.com/encode/starlette/issues/495#issuecomment-494008175). This is the reason why request body data won't be sent to the web UI.
```
Comment (joshuap) [inline README.md:195]:
```suggestion
Consuming the request body in an ASGI application's middleware is [problematic and discouraged](https://github.com/encode/starlette/issues/495#issuecomment-494008175). This is the reason why request body data won't be sent to the web UI.
```

File: README.md [hunk 7]
```diff
@@ -130,6 +130,99 @@ def lambda_handler(event, context):
...
+
+honeybadger.configure(api_key="<your-api-key>")
+app = FastAPI()
+app.router.route_class = contrib.HoneybadgerRoute
+
+router = APIRouter(route_class=contrib.HoneybadgerRoute)
+
+```
+
+### [Starlette](https://www.starlette.io/).
```
Comment (joshuap) [inline README.md:213]:
```suggestion
### Starlette
```

File: README.md [hunk 8]
```diff
@@ -130,6 +130,99 @@ def lambda_handler(event, context):
...
+You can configure Honeybadger to work with Starlette just like in any other ASGI framework.
```
Comment (joshuap) [inline README.md:214]:
```suggestion
You can configure Honeybadger to work with [Starlette](https://www.starlette.io/) just like in any other ASGI framework.
```

**Review by joshuap — COMMENTED** [review]
This looks great to me overall (although I will defer to @Kelvin4664's for code review). I made a few suggestions for the docs. Thank you for submitting this @stefanondisponibile!!

**Comment by joshuap** [comment]
I forgot to mention—we will also need to add a [CHANGELOG.md](https://github.com/honeybadger-io/honeybadger-python/blob/master/CHANGELOG.md) entry before merging this (under the "Unreleased" heading).

**Comment by stefanondisponibile** [comment]
Hey @joshuap, thanks for your edits! I updated the changelog. :+1:

**Comment by joshuap** [comment]
Thanks!! @Kelvin4664 feel free to merge this if you think it's ready.

**Comment by Kelvin4664** [comment]
Unable to install async-asgi-testclient. @stefanondisponibile is there something we should do different?

**Comment by stefanondisponibile** [comment]
Hey @Kelvin4664 , sorry but today I'm just on mobile and don't have a clear idea on why it is failing. I'll have a closer look and let you know my guess!

**Pull request merged**